pages/map.py

```diff
@@ -57,45 +57,78 @@ def create_lines_map(trips_group_by_user_id, user_id_list):
     return fig
 
 
-def create_heatmap_fig(data):
+def get_map_coordinates(trips_group_by_user_mode, user_mode_list):
+    coordinates = {'lat': [], 'lon': [], 'color':[]}
+    for user_mode in user_mode_list:
+        color = trips_group_by_user_mode[user_mode]['color']
+        trips = trips_group_by_user_mode[user_mode]['trips']
+
+        for trip in trips:
+            coordinates['lon'].append(trip['start_coordinates'][0])
+            coordinates['lon'].append(trip['end_coordinates'][0])
+            coordinates['lat'].append(trip['start_coordinates'][1])
+            coordinates['lat'].append(trip['end_coordinates'][1])
+            coordinates['color'].extend([color,color])
+    return coordinates
+
+
+def create_heatmap_fig(trips_group_by_user_mode, user_mode_list):
+    coordinates = get_map_coordinates(trips_group_by_user_mode, user_mode_list)
```

Comment:
Given that you have a consistent pattern across call the `create_XXX_fig`, with the first line being

```
coordinates = get_map_coordinates(trips_group_by_user_mode, user_mode_list)
```

it seems like you can just pass in the coordinates directly. Then you don't have to iterate over all the trips every time the user switches between map types.